File: backend/routes/vouchers.ts
import express, { Router, Request, Response } from 'express';
import db from '../config/database';

const router: Router = express.Router();

interface VoucherRequestBody {
  code: string;
}

// Validate voucher code
router.post('/validate', async (req: Request<{}, {}, VoucherRequestBody>, res: Response) => {
  try {
    const { code } = req.body;
    
    if (!code) {
      return res.status(400).json({ 
        success: false, 
        error: 'Voucher code is required' 
      });
    }
    
    const result = await db.query(
      `SELECT * FROM vouchers 
       WHERE code = $1 
       AND is_active = true 
       AND (expires_at IS NULL OR expires_at > NOW())
       AND (max_uses IS NULL OR used_count < max_uses)`,
      [code.toUpperCase()]
    );
    
    if (result.rows.length === 0) {
      return res.status(404).json({ 
        success: false, 
        error: 'Invalid or expired voucher code' 
      });
    }
    
    const voucher = result.rows[0];
    return res.json({ 
      success: true, 
      voucher: {
        code: voucher.code,
        discountType: voucher.discount_type,
        discountValue: parseFloat(voucher.discount_value),
      }
    });
  } catch (error) {
    const err = error as Error;
    console.error('Error validating voucher:', err);
    return res.status(500).json({ success: false, error: err.message });
  }
});

export default router;

File: backend-golang/internal/handlers/vouchers.go
package handlers

import (
	"context"
	"encoding/json"
	"net/http"

	"inanhxink/backend-golang/internal/config"
)

// POST /api/vouchers/validate
func ValidateVoucher(w http.ResponseWriter, r *http.Request) {
	var body struct {
		Code string `json:"code"`
	}
	if err := json.NewDecoder(r.Body).Decode(&body); err != nil || body.Code == "" {
		BadRequest(w, "Voucher code is required")
		return
	}

	rows, err := config.DB.Query(context.Background(), `
		SELECT code, discount_type, discount_value FROM vouchers
		WHERE UPPER(code) = UPPER($1) AND is_active = true
		AND (expires_at IS NULL OR expires_at > NOW())
		AND (max_uses IS NULL OR used_count < max_uses)`,
		body.Code)
	if err != nil {
		InternalError(w, err)
		return
	}
	row, err := CollectOne(rows)
	if err != nil || row == nil {
		JSON(w, 404, map[string]any{"success": false, "error": "Invalid or expired voucher code"})
		return
	}

	var discountVal float64
	switch v := row["discount_value"].(type) {
	case float64:
		discountVal = v
	case string:
		json.Unmarshal([]byte(v), &discountVal) //nolint
	}

	OK(w, map[string]any{
		"success": true,
		"voucher": map[string]any{
			"code":          row["code"],
			"discountType":  row["discount_type"],
			"discountValue": discountVal,
		},
	})
}
